components/service/firefox-accounts/src/main/java/mozilla/components/service/fxa/Utils.kt

```diff
@@ -4,24 +4,68 @@
 
 package mozilla.components.service.fxa
 
+import mozilla.components.concept.sync.AuthException
+import mozilla.components.concept.sync.AuthExceptionType
+import mozilla.components.service.fxa.manager.authErrorRegistry
+import mozilla.components.support.base.log.logger.Logger
+
 /**
- * Runs a provided lambda, and if that throws non-panic FxA exception, runs the second lambda.
+ * Runs a provided lambda, and if that throws non-panic, non-auth FxA exception, runs [handleErrorBlock].
+ * If that lambda throws an FxA auth exception, notifies [authErrorRegistry], and runs [postHandleAuthErrorBlock].
  *
  * @param block A lambda to execute which mail fail with an [FxaException].
- * @param handleErrorBlock A lambda to execute if [block] fails with a non-panic [FxaException].
+ * @param postHandleAuthErrorBlock A lambda to execute if [block] failed with [FxaUnauthorizedException].
+ * @param handleErrorBlock A lambda to execute if [block] fails with a non-panic, non-auth [FxaException].
  * @return object of type T, as defined by [block].
  */
-fun <T> maybeExceptional(block: () -> T, handleErrorBlock: (e: FxaException) -> T): T {
+fun <T> maybeExceptional(
+    block: () -> T,
+    postHandleAuthErrorBlock: () -> T,
+    handleErrorBlock: (e: FxaException) -> T
+): T {
     return try {
         block()
     } catch (e: FxaException) {
         when (e) {
             is FxaPanicException -> {
                 throw e
             }
+            is FxaUnauthorizedException -> {
+                authErrorRegistry.notifyObservers { onAuthErrorAsync(AuthException(AuthExceptionType.UNAUTHORIZED, e)) }
+                postHandleAuthErrorBlock()
+            }
             else -> {
                 handleErrorBlock(e)
             }
         }
     }
 }
+
+/**
+ * Helper method similar to [maybeExceptional], but a lazy default value could be specified via
+ * the [default] block; it'll be used in case of errors. Execution is wrapped in log statements.
+ */
+fun <T> maybeExceptionalOr(logger: Logger, operation: String, default: () -> T, block: () -> T): T {
```

Comment:
I don't think this method needs a different name. The overload is good enough, no?